src/why3prop.ml

```diff
@@ -0,0 +1,135 @@
+(** Translation of lambdapi goals to why3 goals in propositional case *)
+
+open Terms
+open Extra
+
+exception NoGoalTranslation
+
+(* a map from lambdapi terms to Why3 constants *)
+type cnst_table = (term * Why3.Term.lsymbol) list
+
+(* a map from labels to Why3 terms. *)
+type ctxt_labels = Why3.Term.term StrMap.t
+
+(** [get_newname ()] generates a new axiom name. *)
+let get_newname : unit -> string =
+    (* number of axioms proved whith the Why3 tactic *)
+    let nbr_axioms : int ref = ref 0 in fun () ->
+    nbr_axioms := !nbr_axioms + 1;
+    "Why3axiom_" ^ (string_of_int !nbr_axioms)
+
+(* builtins configuration for propositional logic *)
+type prop_config =
+  { symb_P     : sym (** Encoding of propositions.        *)
+  ; symb_T     : sym (** Encoding of types.               *)
+  ; symb_or    : sym (** Disjunction(∨) symbol.           *)
+  ; symb_and   : sym (** Conjunction(∧) symbol.           *)
+  ; symb_imp   : sym (** Implication(⇒) symbol.           *)
+  ; symb_bot   : sym (** Bot(⊥) symbol.                   *)
+  ; symb_top   : sym (** Top(⊤) symbol.                   *)
+  ; symb_not   : sym (** Not(¬) symbol.                   *) }
+
+(** [get_prop_config pos builtins] set the builtins configuration using
+    [builtins] *)
+let get_prop_config :
+    Pos.popt -> sym StrMap.t -> prop_config = fun pos builtins ->
+    let find_sym key =
+        try StrMap.find key builtins with Not_found ->
+        Console.fatal pos "Builtin symbol [%s] undefined." key
+    in
+    { symb_P     = find_sym "P"
+    ; symb_T     = find_sym "T"
+    ; symb_or    = find_sym "or"
+    ; symb_and   = find_sym "and"
+    ; symb_imp   = find_sym "imp"
+    ; symb_bot   = find_sym "bot"
+    ; symb_top   = find_sym "top"
+    ; symb_not   = find_sym "not" }
+
+(** [translate pos builtins (hs, g)] translates from lambdapi to Why3 goal [g]
+    using the hypothesis [hs]. The function return
+    [constants_table, hypothesis, formula] where :
+    - [constants_table] maps abstracted Lambdapi terms to Why3 constants
+    - [hypothesis] maps abstracted labels to Why3 terms (presentation of [hs])
+    - [formula] Why3 term representing the goal [g].  *)
+
+let rec translate : Pos.popt -> sym StrMap.t -> (Env.env * term) ->
+    cnst_table * ctxt_labels * Why3.Term.term =
+    fun pos builtins  (hs, g) ->
+    let cfg = get_prop_config pos builtins in
+    let (constants_table, hypothesis) =
+        List.fold_left (translate_context cfg) ([], StrMap.empty) hs in
+    try
+        let (constants_table, formula) = t_goal cfg constants_table g in
+        (constants_table, hypothesis, formula)
+    with NoGoalTranslation ->
+        Console.fatal pos "The term [%a] is not of the form [P _]"
+        Print.pp g
+
+(** [translate_context cfg (l_constants, l_hypothesis) (hyp_name, (_, hyp))]
+    translate the context [hyp] with the label [hyp_name] and add it in
+    [l_hypothesis] with the why3 constants [l_constants]. *)
+and translate_context :
```

Comment:
Please rename translate_context into translate_hyp as its argument is not a context by a hypothesis.

Reply:
Done in a29414085332857be96b62ab67a96960b46d746d.